### ISSUE TITLE
Exception: There is no active TOTP setup

### ISSUE THREAD

**Issue by dkliss** [aws-amplify/amplify-flutter#4490]
### Description
-EDIT, Correction on sync

Hi, 

When setting up TOTP, after Amplify.Auth.setUpTotp() has been setup, we will be  at Amplify.Auth.verifyTotpSetup(totpCode). The verifyTotpSetup() only allows **single attempt** to enter correct code. And if second attempt is made I get below exception. Is this how it is expected to work (i.e. only one attempt)?  

```
flutter: Error confirming attribute update: AuthPreconditionException {
  "message": "There is no active TOTP setup",
  "recoverySuggestion": "Call Amplify.Auth.setUpTotp first"
}
```

```
amplify_flutter: ^1.6.1
amplify_auth_cognito: ^1.6.2

```
To reproduce: 
1. run Amplify.Auth.setUpTotp() 
2. Connect with Authenticator.
3. Run Amplify.Auth.verifyTotpSetup(totpCode) with incorrect code --> **Exception: "Code mismatch"** 
4. Run  Amplify.Auth.verifyTotpSetup(totpCode) with correct code--> **Exception: "There is no active TOTP setup."**
5. Re-run Amplify.Auth.setUpTotp(), which will create new code with different sync secret. 
6. Re-scan new qr code with Authenticator.
7. Re-Run Amplify.Auth.verifyTotpSetup(totpCode) with correct code  -> **This function is void and no success  response is received* --> **No exception means success.**  

In  above scenario, even if I do not remove (i.e. execute step 6) authenticator scanned in step 2 above (Connect with Authenticator), Amplify.Auth.verifyTotpSetup() still works even after a new Amplify.Auth.setUpTotp() is generated in step 5. 

To reproduce (Step 6 removed from above scenario): 

1. run Amplify.Auth.setUpTotp() 
2. Connect with Authenticator.
3. Run Amplify.Auth.verifyTotpSetup(totpCode) with incorrect code --> **Exception: "Code mismatch"** 
4. Run  Amplify.Auth.verifyTotpSetup(totpCode) with correct code--> **Exception: "There is no active TOTP setup."**
5. Re-run Amplify.Auth.setUpTotp(), which will create new code with different sync secret. 
6 Re-Run Amplify.Auth.verifyTotpSetup(totpCode) with correct code  -> **This function is void and no success  response is received* --> **No exception means success.**  

Whats happening in this seems like old QR code is still in Amplify and hence when I add verify code, that one is used.  

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Platforms

- [X] iOS
- [ ] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.19.1

### Amplify Flutter Version

1.6.1

### Deployment Method

Amplify CLI

**Comment by Equartey** [comment]
Hi @dkliss, good question, the observed behavior is expected when using `Amplify.Auth.verifyTotpSetup()`.

The correct API to call while onboarding a User to MFA is `Amplify.Auth.confirmSignIn(confirmationValue: totpCode)`. 

If the code is wrong, a `CodeMismatchException` will be thrown allowing you to try again with no need to call `verifyTotpSetup()`. After a valid code, the user will be signed in. 

We have a command line example app with this implementation found on our [github](https://github.com/aws-amplify/amplify-flutter/blob/main/packages/auth/amplify_auth_cognito_dart/example/bin/example.dart#L146-L167).

**Comment by Equartey** [comment]
Hi @dkliss, I looked into it further and it does actually appear to be a bug. Thank you for the extra context with your use case as it helped narrow down the issue.

I'm working on addressing it and will update you here when I can.

**Comment by Equartey** [comment]
Hi @dkliss this has been fixed in `amplify_auth_cognito: ^1.7.1`. 

I'm going to close this issue as complete, please reach out if you encounter another issue.

**Comment by Equartey** [comment]
Hi @dkliss, we catch the exception internally, but allow multiple calls to `verifyTotpSetup` without the need to restart setup. 

I suspect the problem you're facing is not receiving enough feedback during the code verification process. Basically, how would you determine if the attempt was successful or a failure requiring a retry. Which is understandably frustrating.

Can you confirm you're able to submit multiple calls to `verifyTotpSetup` when passing an incorrect code without having to restart setup (your steps 1-3)?

Try: 
1. call `Amplify.Auth.setUpTotp()`
2. call `Amplify.Auth.verifyTotpSetup(INVALID_CODE)` (can be repeated) -- note: no exception thrown
3. call `Amplify.Auth.verifyTotpSetup(VALID_CODE)`
4. Successful setup

Meanwhile, we will investigate how to provide better feedback to you.

**Comment by Equartey** [comment]
Hi @dkliss, I have a work around for you in the meantime.

When setting up a user's TOTP for the first time, you can check the user's MFA preference to see if TOTP is enabled. Now this will only be of value during initial setup on the first device as it only checks to see if TOTP is enabled for the user not, if the current attempt was successful. 

 ```dart
final plugin = Amplify.Auth.getPlugin(AmplifyAuthCognito.pluginKey);
final response = await plugin.fetchMfaPreference();
print(response);
// TOTP not enabled
// UserMfaPreference {
//   "enabled": [],
//   "preferred": null
// }
// TOTP enabled
// UserMfaPreference {
//   "enabled": [
//     "TOTP" // Can check if this exists to determine if setup was successful
//   ],
//   "preferred": "TOTP"
// }
```

Not a perfect fix, but hopefully unblocks you while we investigate a better solution.

**Comment by Equartey** [comment]
Hi @dkliss, I appreciate the discussion so far. It's been helpful for me to understand your use case. I've opened https://github.com/aws-amplify/amplify-flutter/pull/4558 to throw an exception during TOTP setup. Will update you when it's released. 

I believe your updated approach, limiting users to one Authenticator device, improves the user experience until we can handle multiple TOTP secrets properly. Which we will continue to track in the other issue you've opened, https://github.com/aws-amplify/amplify-flutter/issues/4509.

**Comment by Equartey** [comment]
Hi @dkliss, we ended up reverting 1.7.1 as it broke existing behavior, but we have released the idle fix. Please consume  `amplify_auth_cognito: ^1.7.2`.

`Amplify.Auth.verifyTotpSetup(INVALID_CODE)` now throws an `EnableSoftwareTokenMfaException` and allows retries without restarting the setup process. 

Again, thank you for bringing this to our attention. I'm going to close this, but please reach out if you have any issues.
